[Speculative decoding] Add ngram prompt lookup decoding

leiwen83 opened the pull request:
Algo details could refer to this blog post:
https://huggingface.co/blog/assisted-generation

Code directly refer to transformers's current implementation. https://github.com/huggingface/transformers/pull/27775

Since we directly get draft from prompt, there is no need another model or modified model to get the proposal, it would be the most convenient way to enjoy the speedup of speculation.

leiwen83 commented:
Implementaion for the feature of ngram prompt lookup mentioned in https://github.com/vllm-project/vllm/pull/2188

leiwen83 commented:
@cadedaniel 
Could you help to review this PR, especially whether we shall have some more abstraction taking some other speculatiation type like medusa into consideration?

cadedaniel commented:
@comaniac do you have bandwidth to shepherd this PR?

cadedaniel commented on vllm/spec_decode/multi_step_worker.py:
```diff
@@ -11,6 +12,68 @@
 from vllm.worker.worker import Worker
 
 
+class NGramDrafter():
```
Can we break ngram stuff out into its own file? basically, the file `multi_step_worker.py` can be specialized to draft model, then we can have `ngram_worker.py` in its own file.

leiwen83 replied:
Add ngram_worker.py in the latest commit, also I refactor DraftModelTop1Proposer out, and ngram_worker/multi_step_worker current share the same logic.

cadedaniel commented on vllm/spec_decode/spec_decode_worker.py:
```diff
@@ -92,13 +92,13 @@ def __init__(
         self.scorer: SpeculativeScorer
 
-    def init_device(self) -> None:
+    def init_device(self, spec_type: Optional[str] = None) -> None:
```
can we remove this arg and pass it via config / init? we want to keep `init_device` consistent with other implementations of WorkeBase interface

leiwen83 replied:
Done as suggested.

cadedaniel commented on tests/spec_decode/e2e/test_correctness.py:
```diff
@@ -89,6 +89,49 @@ def test_spec_decode_e2e_logical_flow(test_llm_generator, batch_size: int):
+    ])
+@pytest.mark.parametrize("batch_size", [1])
+@pytest.mark.parametrize("seed", [1])
+def test_validate_ngram_prompt_lookup(baseline_llm_generator,
```
I need to write docs still, but FYI these tests are a bit sensitive to batch size / prompts. In my investigation I find that vLLM logits can be unstable depending on some batching condition (even with torch determinism flags enabled). I haven't looked further to identify if it's framework nondeterminism, vLLM nondeterminism, or a spec decode bug. But since it's very stable for the bs/prompts that work, it's OK to rely on this for correctness.

leiwen83 replied:
Got it.

comaniac commented:
Thanks for the PR! I'll review it tomorrow

comaniac commented on vllm/spec_decode/multi_step_worker.py:
```diff
@@ -11,6 +12,68 @@
+
+            input_ids = torch.from_numpy(
+                np.asarray(seq_data.get_token_ids(), np.int64))
+            ngram_tensor = input_ids[-1 * self.ngram_len:]
```
In this implementation we only match n-gram equals to the given size. However, we could reduce n-gram size if no match so that we could have more speculative proposals.

leiwen83 replied:
add the logic for checking window between ngram_prompt_lookup_min and ngram_prompt_lookup_max

comaniac commented on vllm/spec_decode/multi_step_worker.py:
```diff
@@ -11,6 +12,68 @@
+        outputs = []
+        arr = np.array(arr).transpose()
+        for i in range(num_steps):
+            token_probs = torch.zeros(
+                (len(seq_group_metadata_list), self.vocab_size),
+                dtype=torch.float32).to(self.device)
+
+            token_ids = torch.from_numpy(arr[i]).long().to(self.device)
+            for idx, token in enumerate(arr[i]):
+                token_probs[idx][token] = 1
```
I feel that this can be done with tensor ops instead of a for-loop to reduce CPU-GPU data transfer.

leiwen83 replied:
use scatter_ func to accelerate this func

leiwen83 commented on vllm/spec_decode/multi_step_worker.py:
```diff
@@ -11,6 +12,68 @@
+class NGramDrafter():
+
+    def __init__(self, device, vocab_size):
+        self.ngram_len = 2
```
Add --ngram-prompt-lookup-max/--ngram-prompt-lookup-min for this configuration.

leiwen83 commented on vllm/spec_decode/multi_step_worker.py:
```diff
@@ -11,6 +12,68 @@
+            seq_data = next(iter(seq_group_metadata.seq_data.values()))
+
+            input_ids = torch.from_numpy(
+                np.asarray(seq_data.get_token_ids(), np.int64))
```
The loop handles seq_data inside seq_group_metadata_list, I'm not sure whether we could do it in batch as different seq_data may have diffent prompt match exit point...

comaniac commented on vllm/spec_decode/ngram_worker.py:
```diff
@@ -0,0 +1,260 @@
+        self._raise_if_unsupported(seq_group_metadata_list, blocks_to_swap_in,
+                                   blocks_to_swap_out, blocks_to_copy)
+
+        # Shallow copy input data so modifications (such as appending tokens)
+        # do not cause side-effects.
+        copied_seq_group_metadata_list = self._shallow_copy_inputs(
+            seq_group_metadata_list)
```
IIUC, appending speculative tokens is a general step for every speculative decoding algorithms. @cadedaniel could you help confirm whether we need to have this shallow copy or not? Thanks.

cadedaniel replied:
yeah, we should remove all this multi step stuff for ngram

cadedaniel replied:
I believe we don't need modification to seq_group_metadata_list, which is the only motivation for the shallow copy. Can we live without it in the ngram worker?

leiwen83 replied:
Would apply it in latest commit~

comaniac commented on vllm/spec_decode/ngram_worker.py:
```diff
@@ -0,0 +1,260 @@
+
+        return new_seq_group_metadata_list
+
+    def _assert_enough_kv_space(
```
I didn't find this function being used anywhere?

leiwen83 replied:
would remove it

comaniac reviewed (commented):
Another batch of comments. Also we would need unit tests for this PR.

cadedaniel commented on vllm/spec_decode/ngram_worker.py:
```diff
@@ -0,0 +1,260 @@
+from vllm.worker.worker import Worker
+
+
+class NGramWorker(Worker):
```
Let's make this inherit from WorkerBase, not Worker. Unless I miss something we don't need anything from the Worker class.

leiwen83 replied:
I try use the WorkerBase here, but got below error:
TypeError: Can't instantiate abstract class NGramWorker with abstract methods add_lora, determine_num_available_blocks, execute_model, get_cache_block_size_bytes, initialize_cache, list_loras, remove_lora

cadedaniel replied:
> Got it, shall we also need make multi_step_worker inherit from WorkerBase too?

multi_step_worker should ideally inherit from WorkerBase, but it requires a larger refactor. You can help with that if you want :) the TL;DR is to decouple the sampler from CPU serialization of python objects. If we do this then the multi step worker can inherit directly WorkerBase and become decoupled from Worker. But it's a lot of work right now, with the only benefit being cleaner software design. We'll do it soon as it will also improve performance, but not now.

> I try use the WorkerBase here, but got below error:
TypeError: Can't instantiate abstract class NGramWorker with abstract methods add_lora, determine_num_available_blocks, execute_model, get_cache_block_size_bytes, initialize_cache, list_loras, remove_lora

You can inherit from [`LoraNotSupportedWorkerBase`](https://github.com/vllm-project/vllm/blob/1e8f4252aa163041094a8fedbb701a98d7087d7c/vllm/worker/worker_base.py#L81-L93) to handle the lora methods. `determine_num_available_blocks` can be a noop; [the proposer worker's `determine_num_available_blocks` is unused in spec decode](https://github.com/vllm-project/vllm/blob/1e8f4252aa163041094a8fedbb701a98d7087d7c/vllm/spec_decode/spec_decode_worker.py#L140-L159). For `get_cache_block_size_bytes` you can return 0, since there is no on-device KV cache for ngram. Initialize_cache can be noop as well (just `pass`), unless there's on-device state you want to initialize.

leiwen83 replied:
Got it, would apply it in latest commit

cadedaniel commented on vllm/spec_decode/ngram_worker.py:
```diff
@@ -0,0 +1,260 @@
+    def init_device(self):
+        super().init_device()
+        # Since ngram worker use no model, fake load_model/execute_model
+        self.load_model = self.fake_load_model
+        self.execute_model = self.fake_execute_model
```
Yeah, I suggest implementing the methods and just have them `pass`. No need for dynamic override.

leiwen83 replied:
Great, it get clearer after this change. I'll apply it in latest commit

cadedaniel commented on vllm/spec_decode/ngram_worker.py:
```diff
@@ -0,0 +1,260 @@
+        )
+
+    @torch.inference_mode()
+    def execute_model_multi_step(
```
The current implementation couples the NGramWorker with the draft model worker a lot. In the future we will have RAG speculation, Medusa, Eagle, that all can use top-1 proposals. Can you look at how we can decouple Top1Proposal logic from both the draft model and ngram, so we can use it for the other ones in the future?

It will also simplify the ngram worker implementation; we won't need this notion of `execute_model_multi_step` and its baggage (shallow copy inputs, check KV size, append new tokens).

cadedaniel replied:
I'm happy to meet to discuss this item further -- it will simplify things a lot in the future (cc @comaniac ).

leiwen83 replied:
I currently move top-1 proposals into top1_proposer.py, and move shallow_copy_inputs into util.py, so that there is no [check KV size, append new tokens] logic for ngram now. 

For shallow_copy_inputs itself, it seems we may remove it from ngram, as it is true we don't need to do the append token stuff.

cadedaniel replied:
+1 to removing shallow_copy_inputs / check KV size / append new tokens

> I currently move top-1 proposals into top1_proposer.py

The issue is the top1 proposer is still couples with the draft model -- it calls `execute_model_multi_step` and also relies on `max_model_len`. Top1Proposing should work for all proposal methods (ngram/eagle/medusa/draft), so before we can call your factor `Top1Proposer` we need to move out the draft-model specific components. As you have it, it's still a `DraftModelTop1Proposer`.

leiwen83 commented on vllm/spec_decode/ngram_worker.py:
```diff
@@ -0,0 +1,260 @@
+
+                    match_found = True
+                    break
+
+            if not match_found:
+                res = [0] * num_steps
```
Got it, would apply it~

leiwen83 commented on vllm/spec_decode/ngram_worker.py:
```diff
@@ -0,0 +1,260 @@
+        outputs = []
+        arr = np.array(arr).transpose()
+        for i in range(num_steps):
+            token_probs = torch.zeros(
+                (len(seq_group_metadata_list), self.vocab_size),
+                dtype=torch.float32).to(self.device)
```
Got it, and token_probs here could be move out of for loop. Test shows it would helps the performance a lot, would apply it in latest commit~

leiwen83 commented on vllm/spec_decode/ngram_worker.py:
```diff
@@ -0,0 +1,260 @@
+        for seq_group_metadata in copied_seq_group_metadata_list:
+            seq_data = next(iter(seq_group_metadata.seq_data.values()))
+
+            input_ids = torch.from_numpy(
+                np.asarray(seq_data.get_token_ids(), np.int64))
```
It's reasonable, would apply it. Thx~

leiwen83 commented:
> Another batch of comments. Also we would need unit tests for this PR.

I current have one e2e unittest in tests/spec_decode/e2e/test_correctness.py. Do you mean may need some other unit tests?

leiwen83 commented:
@cadedaniel @comaniac ,
Since https://github.com/vllm-project/vllm/pull/3951 is merged, and there is some changes have been made for spec infer sub system, shall we squash this commit and rebase over latest code for easy further code review?

cadedaniel commented:
> @cadedaniel @comaniac ,
Since https://github.com/vllm-project/vllm/pull/3951 is merged, and there is some changes have been made for spec infer sub system, shall we squash this commit and rebase over latest code for easy further code review?
And if squash/rebase is needed, whether I shall force update in this thread or open another PR?

You're free to use what strategy you prefer to remove the merge conflicts. I often will merge main into my dev branch but there's tradeoffs. I'd keep this PR open, you can push/force push as necessary!

And sorry for the conflicts -- PR 7/9 was the last large one to the subsystem. I expect less merge conflicts going forward.

> >I current have one e2e unittest in tests/spec_decode/e2e/test_correctness.py. Do you mean may need some other unit tests?
>
> Sorry let me make it more clear. The unit tests should cover as many cases as possible. For example, batch size > 1 with some seqs find a match but others don't; n-gram size from 1 to 3; long/short speculative sizes, etc.

+1. We'll want the following tests:
* Batch size 1 greedy equality
* Batch size >1 greedy equality
* A test that covers when there's no ngram match for any sequence
* A test that covers when there's ngram matches for some sequences in a batch, but not all
* A test that covers when there's ngram matches for all seqs in the batch (can omit since bs>1 greedy equality will likely cover it)
* Test various ngram sizes / speculative sizes
* Test greedy equality under preemption.

Most of these can be copied from [the existing tests](https://github.com/vllm-project/vllm/blob/main/tests/spec_decode/e2e/test_correctness.py) and refitted for ngram speculation; I suggest making a new file for spec ngram test correctness (so the current one can be draft model).

leiwen83 commented:
@cadedaniel @comaniac 
I squash and rebase over latest main branch, while fix all issues so far and add the necessary test cases, would you mind to take a look at the refreshed PR?

And @cadedaniel, I make a change for the meta-llama/Llama-2-7b-chat-hf in e2e test to NousResearch/Llama-2-7b-chat-hf, since meta-llama repo is an restricted repo which need apply access request, which is not easy, I took for months but still fail to get the access right, and I think many people like me may face the similar issue. If you feels it is unnecessary, I could change it back to meta-llama one.

Thx~

comaniac commented on tests/spec_decode/e2e/test_ngram_correctness.py:
```diff
@@ -0,0 +1,285 @@
+            "ngram_prompt_lookup_max": 3,
+        }
+        # Try a range of common k, as well as large speculation.
+        for k in [1, 2, 3, 4, 5, 6, 7, 8, 9, 63]
```
We may just test `[1,3,7,10,63]` to reduce test time.

leiwen83 replied:
Got it, would apply~

comaniac commented on vllm/spec_decode/ngram_worker.py:
```diff
@@ -0,0 +1,188 @@
+                    res = res[match_indices[0] + ngram_size:match_indices[0] +
+                              ngram_size + sample_len]
+                    res_len = len(res)
+                    # pad 0 towards output as sample_len tokens required
+                    for i in range(res_len, sample_len):
+                        res.append(0)
```
```suggestion
                    # pad 0 towards output as sample_len tokens required
                    res += [0] * (sample_len - res_len)
```

leiwen83 replied:
It looks better, thx~

comaniac reviewed (commented):
Otherwise LGTM (the n-gram worker). I'll leave the rest to @cadedaniel

leiwen83 commented on tests/spec_decode/e2e/test_compatibility.py:
```diff
@@ -90,7 +90,7 @@ def test_spec_decode_xfail_chunked_prefill(test_llm_generator):
 @pytest.mark.parametrize(
     "common_llm_kwargs",
     [{
-        "model": "meta-llama/Llama-2-7b-chat-hf",
```
change it back in latest commit

leiwen83 commented on tests/spec_decode/e2e/test_ngram_correctness.py:
```diff
@@ -0,0 +1,285 @@
+    ])
+@pytest.mark.parametrize("batch_size", [64])
+@pytest.mark.parametrize("seed", [1])
+def test_spec_decode_e2e_greedy_correctness_tiny_model_large_bs(
```
merged it as suggested

cadedaniel commented:
will take another pass today

cadedaniel reviewed (approved):
Just minor comments, otherwise LGTM! Can you address/respond and then we'll merge.

wenlei03 committed:
[Speculative decoding] Add ngram prompt lookup decoding

Algo details could refer to this blog post:
https://huggingface.co/blog/assisted-generation

Code directly refer to transformers's current implementation.
https://github.com/huggingface/transformers/pull/27775

Since we directly get draft from prompt, there is no need
another model or modified model to get the proposal, it would be
the most convenient way to enjoy the speedup of speculation.

wenlei03 committed:
fix spell

leiwen83 commented:
@cadedaniel all note has been addressed, and rebase it against latest code, could you take another look?

cadedaniel commented:
Looks good, thanks @leiwen83 ! Thanks for contributing to vLLM 😃
